## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [10]:
# View the column names in the data
column_names = df.columns
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [12]:
# Use the describe function to gather some basic statistics
basic_stats = df.describe()
print(basic_stats)



          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [13]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.head(10)


first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  
0            wall art     1096.80     762.71         7.50  105            1  
1                pens       24.95      15.09         1.49   21            0  
2  project management       13.52       7.86         1.68   39            6  
3                pens       36.42      24.85         1.23   29            3  
4                misc      195.10     108.17        46.43   20            1  
5              plants      172.84     154.44         2.79   78            0  
6    kitchen supplies      245.56     231.95         1.73   56            5  
7                pens       20.99      16.97         5.12   22            4  
8      paper products       83.86      59.00         0.27   49            3  
9             seating       66.88      54.15         2.47   78            1

In [16]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().nlargest(3)
print(top_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [18]:
# For the category with the most entries,
# which subcategory had the most entries?
top_category = top_categories.index[0]
top_subcategory = df[df['category'] == top_category]['subcategory'].value_counts().idxmax()
print(top_subcategory)

bathroom supplies


In [19]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().nlargest(5)
print(top_clients)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [20]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = top_clients.index.tolist()


In [22]:
# How many total units (the qty column) did the
# client with the most entries order order?
total_units = df[df['client_id'] == top_client_ids[0]]['qty'].sum()
print(total_units)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [24]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']
print(df.head())

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor  

In [35]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
df['shipping_price'] = df['unit_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
print(df.head())

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.5

In [36]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['total_price'] = df['subtotal'] + df['shipping_price'] + (df['subtotal'] + df['shipping_price']) * 0.0925
print(df.head())


       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.5

In [37]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
print(df.head())

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.5

In [38]:
# Create a column for the profit of
# each line using line cost and line price

df['profit'] = df['total_price'] - df['cost']
print(df.head())

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.5

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.
# I exported the data into a .csv file to pull into Excel to determine where the errors were, analyzed the data "output2.csv" which I'm including in my repo for reference, and by grouping by order ID (not client ID) which is what the instructions say, I get different values than this (commented below and provided in README file)
Order ID 2742071 had a total price of \$152,811.89
# My Order ID 2742071 had a total price of \$139,844.51

Order ID 2173913 had a total price of \$162,388.71
# My Order ID 2742071 had a total price of \$119,838.36

Order ID 6128929 had a total price of \$923,441.25
# My Order ID 6128929 had a total price of \$75,2748.75


In [45]:
# Check your work using the totals above
# Group by 'order_id' and sum 'total_price'
order_totals = df.groupby('order_id')['total_price'].sum()

order_ids_to_check = [2742071, 2173913, 6128929]
order_totals_check = order_totals[order_totals.index.isin(order_ids_to_check)]
order_totals_check



order_id
2173913    119838.357050
2742071    139844.512025
6128929    752748.745475
Name: total_price, dtype: float64

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [78]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_client_totals = df.groupby('client_id')['total_price'].sum()
top_client_totals[top_client_ids]
print(top_client_totals[top_client_ids])


KeyError: 'client_id'

In [77]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

Summary_df = pd.DataFrame({top_client_ids[0]: [df[df['client_id'] == top_client_ids[0]]['qty'].sum(), df[df['client_id'] == top_client_ids[0]]['shipping_price'].sum(), df[df['client_id'] == top_client_ids[0]]['total_price'].sum(), df[df['client_id'] == top_client_ids[0]]['profit'].sum()],
                            top_client_ids[1]: [df[df['client_id'] == top_client_ids[1]]['qty'].sum(), df[df['client_id'] == top_client_ids[1]]['shipping_price'].sum(), df[df['client_id'] == top_client_ids[1]]['total_price'].sum(), df[df['client_id'] == top_client_ids[1]]['profit'].sum()],
                            top_client_ids[2]: [df[df['client_id'] == top_client_ids[2]]['qty'].sum(), df[df['client_id'] == top_client_ids[2]]['shipping_price'].sum(), df[df['client_id'] == top_client_ids[2]]['total_price'].sum(), df[df['client_id'] == top_client_ids[2]]['profit'].sum()],
                            top_client_ids[3]: [df[df['client_id'] == top_client_ids[3]]['qty'].sum(), df[df['client_id'] == top_client_ids[3]]['shipping_price'].sum(), df[df['client_id'] == top_client_ids[3]]['total_price'].sum(), df[df['client_id'] == top_client_ids[3]]['profit'].sum()],
                            top_client_ids[4]: [df[df['client_id'] == top_client_ids[4]]['qty'].sum(), df[df['client_id'] == top_client_ids[4]]['shipping_price'].sum(), df[df['client_id'] == top_client_ids[4]]['total_price'].sum(), df[df['client_id'] == top_client_ids[4]]['profit'].sum()]},
                            index=['Total Units Purchased', 'Total Shipping Price', 'Total Revenue', 'Total Profit'])
Summary_df = Summary_df.transpose()
Summary_df.sort_values(by=['Total Profit'], ascending=False, inplace=True)
print(Summary_df)


KeyError: 'client_id'

In [72]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
df = df.rename(columns={'first': 'First Name', 'last': 'Last Name', 'job': 'Job', 'phone': 'Phone', 'email': 'Email', 'client_id': 'Client ID', 'order_id': 'Order ID', 'order_date': 'Order Date', 'category': 'Category', 'subcategory': 'Subcategory', 'product': 'Product', 'qty': 'Quantity', 'unit_price': 'Unit Price', 'unit_cost': 'Unit Cost', 'unit_weight': 'Unit Weight', 'order_week': 'Order Week', 'order_year': 'Order Year', 'line_number': 'Line Number', 'subtotal': 'Subtotal', 'shipping_price': 'Shipping Price', 'total_price': 'Total Revenue (USD Millions)', 'cost': 'Cost', 'profit': 'Total Profit (USD)'})



print(df.head())

  First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year  ... Unit Price Unit Cost Unit Weight  Quantity  \
0          17        2023  ...    1096.80    762.71      

In [75]:
# Sort the updated data by "Total Profit" form highest to lowest
sorted_df = df.sort_values(by=['Profit (USD)'], ascending=False)

KeyError: 'Total Profit (USD)'